### PR TITLE
chickenPackages: Introduce overrides

### PR DESCRIPTION
**Pull request by nagy**
###### Description of changes

This fixes around half of the broken chicken builds.

cc @konst-aa @wegank @marsam

@corngood  if you are still interested.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/chicken/5/overrides.nix [hunk 1]
```diff
@@ -1,2 +1,46 @@
...
+      ++ (if lib.isList pkg then pkg else [ pkg ]);
+  };
+  addToBuildInputsWithPkgConfig = pkg: old:
+    (addToBuildInputs pkg old) // {
+      nativeBuildInputs = (old.nativeBuildInputs or [ ]) ++ [ pkgs.pkg-config ];
+    };
+in {
+  breadline = addToBuildInputs pkgs.readline;
+  cmark = addToBuildInputs pkgs.cmark;
+  epoxy = addToBuildInputsWithPkgConfig pkgs.epoxy;
```
Comment (wegank) [inline pkgs/development/compilers/chicken/5/overrides.nix:14]:
```suggestion
  epoxy = addToBuildInputsWithPkgConfig pkgs.libepoxy;
```

**Comment by wegank** [comment]
```
12 packages failed to build:
chickenPackages.chickenEggs.awful-sqlite3 chickenPackages.chickenEggs.awful-ssl chickenPackages.chickenEggs.dust chickenPackages.chickenEggs.gl-type chickenPackages.chickenEggs.gl-utils chickenPackages.chickenEggs.glls chickenPackages.chickenEggs.hypergiant chickenPackages.chickenEggs.noise chickenPackages.chickenEggs.sdl2-image chickenPackages.chickenEggs.sdl2-ttf chickenPackages.chickenEggs.sourcehut chickenPackages.chickenEggs.topham

27 packages built:
chickenPackages.chickenEggs.breadline chickenPackages.chickenEggs.cairo chickenPackages.chickenEggs.cmark chickenPackages.chickenEggs.epoxy chickenPackages.chickenEggs.exif chickenPackages.chickenEggs.expat chickenPackages.chickenEggs.gemini chickenPackages.chickenEggs.gemini-client chickenPackages.chickenEggs.glfw3 chickenPackages.chickenEggs.imlib2 chickenPackages.chickenEggs.magic chickenPackages.chickenEggs.mdh chickenPackages.chickenEggs.nanomsg chickenPackages.chickenEggs.openssl chickenPackages.chickenEggs.rocksdb chickenPackages.chickenEggs.sdl-base chickenPackages.chickenEggs.sdl2 chickenPackages.chickenEggs.sqlite3 chickenPackages.chickenEggs.stemmer chickenPackages.chickenEggs.stfl chickenPackages.chickenEggs.taglib chickenPackages.chickenEggs.uuid-lib chickenPackages.chickenEggs.ws-client chickenPackages.chickenEggs.yaml chickenPackages.chickenEggs.zlib chickenPackages.chickenEggs.zmq chickenPackages.chickenEggs.zstd
```

**Comment by nagy** [comment]
> 12 packages failed to build:

There should be more failing. For me it is around 85 that still fail. I will mark those as broken in the next PR.

**Comment by wegank** [comment]
> There should be more failing. For me it is around 85 that still fail. I will mark those as broken in the next PR.

I guess other broken packages are not affected by this PR, so `nixpkgs-review` didn't count them.

**Comment by wegank** [comment]
```bash
error: builder for '/nix/store/nh85dbk0z0gs066w6zh6g5dznz0v7jn9-chicken-gl-utils-0.8.0.drv' failed with exit code 70;
       last 10 log lines:
       > /build/gl-utils-0.8.0/gl-utils-core.c:13:10: fatal error: epoxy/gl.h: No such file or directory
       >    13 | #include <epoxy/gl.h>
       >       |          ^~~~~~~~~~~~
       > compilation terminated.
       >
       > Error: shell command terminated with non-zero exit status 256: 'gcc' '/build/gl-utils-0.8.0/gl-utils-core.c' -o '/build/gl-utils-0.8.0/gl-utils-core.o' -c  -fno-strict-aliasing -fwrapv -DHAV
E_CHICKEN_CONFIG_H -DC_ENABLE_PTABLES -Os -fomit-frame-pointer -fPIC -DPIC -DC_SHARED -I/build/gl-utils-0.8.0 -I/nix/store/4khq6crr0vmrk73mp92pccnbxfwh4i50-chicken-5.3.0/include/chicken
       >
       > Error: shell command terminated with nonzero exit code
       > 256
       > "sh /build/gl-utils-0.8.0/gl-utils.build.sh"
       For full logs, run 'nix log /nix/store/nh85dbk0z0gs066w6zh6g5dznz0v7jn9-chicken-gl-utils-0.8.0.drv'.
error: 2 dependencies of derivation '/nix/store/3yc1n7sszbnnm6v1i397119wggkdy0x1-chicken-gl-type-0.4.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/aa62qsn953mwxgvbnjm2r9v6kq2xah1r-chicken-glls-0.12.2.drv' failed to build
error: 1 dependencies of derivation '/nix/store/ah2vv8yjx2y3iy48cq3sn2ysqdif3kd4-chicken-noise-0.2.0.drv' failed to build
error: 6 dependencies of derivation '/nix/store/rnb3nnz275zryj40yhg48clschdssh9p-chicken-hypergiant-0.5.3.drv' failed to build
```

Should we put some libs in `propagatedBuildInputs`?

**Comment by nagy** [comment]
I have added `libepoxy` as a propagated input. That has fixed the `gl-utils` egg at least.

File: pkgs/development/compilers/chicken/5/overrides.nix [hunk 2]
```diff
@@ -1,2 +1,51 @@
...
+  expat = addToBuildInputs pkgs.expat;
+  glfw3 = addToBuildInputsWithPkgConfig pkgs.glfw3;
+  imlib2 = addToBuildInputsWithPkgConfig pkgs.imlib2;
+  magic = addToBuildInputs pkgs.file;
+  mdh = addToBuildInputs pkgs.pcre;
+  nanomsg = addToBuildInputs pkgs.nanomsg;
+  openssl = addToBuildInputs pkgs.openssl;
+  rocksdb = addToBuildInputs pkgs.rocksdb;
+  sdl-base = addToBuildInputs pkgs.SDL;
+  sdl2 = addToBuildInputs pkgs.SDL2;
```
Comment (wegank) [inline pkgs/development/compilers/chicken/5/overrides.nix:30]: I guess something like this should also work:

```suggestion
  sdl2 = addToPropagatedBuildInputs pkgs.SDL2;
  sdl2-image = addToBuildInputs pkgs.SDL2_image;
  sdl2-ttf = addToBuildInputs pkgs.SDL2_ttf;
```

File: pkgs/development/compilers/chicken/5/overrides.nix [hunk 3]
```diff
@@ -1,2 +1,51 @@
...
+{ pkgs, lib, chickenEggs }:
+let
+  addToBuildInputs = pkg: old: {
+    buildInputs = (old.buildInputs or [ ])
+      ++ (if lib.isList pkg then pkg else [ pkg ]);
+  };
+  addToBuildInputsWithPkgConfig = pkg: old:
+    (addToBuildInputs pkg old) // {
+      nativeBuildInputs = (old.nativeBuildInputs or [ ]) ++ [ pkgs.pkg-config ];
+    };
```
Comment (wegank) [inline pkgs/development/compilers/chicken/5/overrides.nix:10]: Can we break this up to `pkg: old: addPkgConfig (addToPropagatedBuildInputs pkg old)`?

File: pkgs/development/compilers/chicken/5/overrides.nix (hunk 2 above)
Comment (nagy) [reply pkgs/development/compilers/chicken/5/overrides.nix]: Good idea! I have tried fixing sdl earlier, but couldn't get it to work. That works now.

File: pkgs/development/compilers/chicken/5/overrides.nix (hunk 3 above)
Comment (nagy) [reply pkgs/development/compilers/chicken/5/overrides.nix]: done

File: pkgs/development/compilers/chicken/5/overrides.nix [hunk 4]
```diff
@@ -1,2 +1,56 @@
...
+  zlib = addToBuildInputs pkgs.zlib;
+  zmq = addToBuildInputs pkgs.zeromq;
+  zstd = addToBuildInputs pkgs.zstd;
+
+  cairo = old:
+    (addToBuildInputsWithPkgConfig pkgs.cairo old) // {
+      # missing dependency declaration in egg
+      propagatedBuildInputs = (old.propagatedBuildInputs or [ ])
+        ++ (with chickenEggs; [ srfi-1 srfi-13 ]);
+    };
```
Comment (wegank) [inline pkgs/development/compilers/chicken/5/overrides.nix:52]:
```suggestion
  cairo = old:
    (addToBuildInputsWithPkgConfig pkgs.cairo old)
    // (addToPropagatedBuildInputs (with chickenEggs; [ srfi-1 srfi-13 ]) old);
```

**Comment by wegank** [comment]
cc @marsam

**Review by konst-aa — APPROVED** [review]

**Pull request merged**